### PR TITLE
Increment 5

### PR DESCRIPTION
**Pull request by e1m0re**

### CODE REVIEW

File: cmd/agent/flags.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	"strconv"
+)
+
+type Options struct {
+	flagRunAddr    string
+	reportInterval uint
+	pollInterval   uint
+}
+
+var agentOptions = new(Options)
```
Comment (arttet) [inline cmd/agent/flags.go]:
```suggestion
var agentOptions = &Options{}
```

Comment (arttet) [inline cmd/agent/flags.go]: CHANGE REQUEST: please avoid global variables

File: cmd/agent/flags.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+func parseFlags() {
+	var defaultA = "localhost:8080"
```
Comment (arttet) [inline cmd/agent/flags.go]: CHANGE REQUEST: this is a magic name

File: cmd/agent/flags.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	if envRunAddr := os.Getenv("ADDRESS"); envRunAddr != "" {
+		defaultA = envRunAddr
+	}
+	flag.StringVar(&agentOptions.flagRunAddr, "a", defaultA, "address and port to run server")
+
+	var defaultR = uint(10)
```
Comment (arttet) [inline cmd/agent/flags.go]:
```suggestion
	defaultR := uint(10)
```

File: cmd/agent/main.go [hunk 4]
```diff
@@ -1,3 +1,101 @@
...
+	"math/rand"
+	"net/http"
+	"net/url"
+	"runtime"
+	"time"
+
+	"github.com/e1m0re/grdn/internal/storage"
+)
+
+func UpdateMetrics(data *storage.MetricsState) {
```
Comment (arttet) [inline cmd/agent/main.go]: CHANGE REQUEST: please split your application by layers. Here is business logic.

File: cmd/server/main.go [hunk 5]
```diff
@@ -1,3 +1,85 @@
...
+	"strconv"
+
+	"github.com/go-chi/chi/v5"
+
+	"github.com/e1m0re/grdn/internal/storage"
+)
+
+var store = storage.NewMemStorage()
+
+func updateMetricHandler(response http.ResponseWriter, request *http.Request) {
```
Comment (arttet) [inline cmd/server/main.go]: CHANGE REQUEST: it should be a method

**Review by arttet — CHANGES_REQUESTED** [review]
Please add
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: cmd/agent/flags.go (hunk 1 above)
Comment (e1m0re) [reply cmd/agent/flags.go]: done

Comment (e1m0re) [reply cmd/agent/flags.go]: ok

File: cmd/agent/flags.go (hunk 2 above)
Comment (e1m0re) [reply cmd/agent/flags.go]: renamed

File: cmd/agent/flags.go (hunk 3 above)
Comment (e1m0re) [reply cmd/agent/flags.go]: done

File: cmd/agent/main.go (hunk 4 above)
Comment (e1m0re) [reply cmd/agent/main.go]: done

File: cmd/server/main.go (hunk 5 above)
Comment (e1m0re) [reply cmd/server/main.go:14]: I've moved Business logic to store.

File: cmd/server/main.go [hunk 6]
```diff
@@ -1,3 +1,69 @@
...
+import (
+	"fmt"
+	"github.com/go-chi/chi/v5"
+	"log"
+	"net/http"
+
+	"github.com/e1m0re/grdn/internal/storage"
+)
+
+var store = storage.NewMemStorage()
```
Comment (arttet) [inline cmd/server/main.go]: CHANGE REQUEST: please don't use global variables

File: cmd/server/main.go [hunk 7]
```diff
@@ -1,3 +1,69 @@
...
+
+func updateMetricHandler(response http.ResponseWriter, request *http.Request) {
```
Comment (arttet) [inline cmd/server/main.go]:
```suggestion
func (h *Handler) updateMetricHandler(response http.ResponseWriter, request *http.Request) {
```

**Review by arttet — CHANGES_REQUESTED** [review]

File: cmd/server/main.go (hunk 6 above)
Comment (e1m0re) [reply cmd/server/main.go]: done

File: cmd/server/main.go (hunk 5 above)
Comment (e1m0re) [reply cmd/server/main.go]: done

File: cmd/server/main.go (hunk 7 above)
Comment (e1m0re) [reply cmd/server/main.go]: Now I understand what you meant and I fixed it.

**Comment by e1m0re** [comment]
done

File: .github/workflows/coveragetest.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.21'
+      - name: Run coverage
+        run: go test -race -coverprofile=coverage.out -covermode=atomic ./...
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v3
```
Comment (arttet) [inline .github/workflows/coveragetest.yml]:
```suggestion
        uses: codecov/codecov-action@v4
```

File: .github/workflows/coveragetest.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
+      - uses: actions/setup-go@v2
+        with:
+          go-version: '1.21'
```
Comment (arttet) [inline .github/workflows/coveragetest.yml]:
```suggestion
          go-version: '1.22'
```

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,8 @@
+test:
+	go test -race -covermode=atomic -coverprofile=coverage.out ./...
+
+run-agent:
+	go run cmd/agent/main.go cmd/agent/flags.go
```
Comment (arttet) [inline Makefile]:
```suggestion
	go run cmd/agent/main.go
```

File: cmd/agent/flags.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+	"strconv"
+)
+
+type Options struct {
+	flagRunAddr    string
+	reportInterval uint
+	pollInterval   uint
+}
+
+var agentOptions = &Options{}
```
Comment (arttet) [inline cmd/agent/flags.go]: CHANGE REQUEST: please avoid global variables

File: internal/srvhandler/srvhandler.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+	"net/http"
+
+	"github.com/go-chi/chi/v5"
+
+	"github.com/e1m0re/grdn/internal/storage"
+)
+
+type Handler struct{}
+
+func (h *Handler) UpdateMetric(store *storage.MemStorage) http.HandlerFunc {
```
Comment (arttet) [inline internal/srvhandler/srvhandler.go]:
```suggestion
func (h *Handler) UpdateMetric(response http.ResponseWriter, request *http.Request) {
```

**Review by arttet — CHANGES_REQUESTED** [review]

**Comment by arttet** [comment]
[2024-03-14T06:49:31.691Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Could not find a repository, try using repo upload token', code='not_found')}

File: cmd/agent/main.go [hunk 13]
```diff
@@ -1,3 +1,58 @@
...
+	"time"
+
+	"github.com/e1m0re/grdn/internal/apiclient"
+	"github.com/e1m0re/grdn/internal/monitor"
+)
+
+func main() {
+	var flagRunAddr string
+
+	flag.StringVar(&flagRunAddr, "a", "localhost:8080", "address and port to run server")
```
Comment (arttet) [inline cmd/agent/main.go:16]: SUGGESTION: it should be in `flags.go`

**Review by arttet — APPROVED** [review]

**Pull request merged**